File: BeyTail_MetalEffects_Port/Reference/Kotlin/GenericGLEffect.kt
package com.beyblade.trailfilter.gl.effects

import android.opengl.GLES20
import com.beyblade.trailfilter.effects.EffectType
import com.beyblade.trailfilter.effects.TrailPoint
import com.beyblade.trailfilter.gl.GLEffect
import com.beyblade.trailfilter.gl.GLHelper
import com.beyblade.trailfilter.gl.GLRenderContext

/**
 * Generic ribbon trail for all effects that don't have custom GL logic.
 * Two ribbon passes: wide glow layer + narrow bright core.
 * Color = beyblade's detected dominant color (pts.last().first.color).
 */
class GenericGLEffect : GLEffect() {

    private var program  = 0
    private var posLoc   = -1
    private var colorLoc = -1

    override fun onGLReady(ctx: GLRenderContext) {
        program  = GLHelper.buildProgram(TRAIL_VERT, TRAIL_FRAG)
        posLoc   = GLES20.glGetAttribLocation(program, "aPosition")
        colorLoc = GLES20.glGetAttribLocation(program, "aColor")
    }

    override fun draw(
        trackData: Map<Int, List<Pair<TrailPoint, Float>>>,
        ctx: GLRenderContext,
        effectType: EffectType
    ) {
        GLES20.glUseProgram(program)
        GLES20.glEnableVertexAttribArray(posLoc)
        GLES20.glEnableVertexAttribArray(colorLoc)
        for ((_, pts) in trackData) {
            if (pts.size < 2) continue
            // colorOverride wins (e.g. fire = red, lightning = yellow); otherwise beyblade color
            val color = effectType.colorOverride ?: pts.last().first.color
            drawRibbon(pts, 0.070f * effectType.glowWidthMult, alphaScale = 0.45f, coreBoost = 0.00f, color, ctx)
            drawRibbon(pts, 0.022f * effectType.coreWidthMult, alphaScale = 0.92f, coreBoost = 0.55f, color, ctx)
        }
        GLES20.glDisableVertexAttribArray(posLoc)
        GLES20.glDisableVertexAttribArray(colorLoc)
    }

    private fun drawRibbon(
        pts: List<Pair<TrailPoint, Float>>,
        halfWidth: Float,
        alphaScale: Float,
        coreBoost: Float,
        headColor: Int,
        ctx: GLRenderContext
    ) {
        val n = pts.size
        if (n < 2) return
        if (n * 2 * BYTES_PER_VERT > ctx.ribbonBuf.capacity()) return

        ctx.ribbonFloats.clear()

        for (i in 0 until n) {
            val (tp, alpha) = pts[i]
            val x = (tp.center.x * 2f - 1f) * ctx.quadScaleX
            val y = (1f - tp.center.y * 2f) * ctx.quadScaleY

            val (nx, ny) = when {
                i == 0 -> {
                    val (tp1, _) = pts[1]
                    GLHelper.segNormal(x, y, tp1.center.x * 2f - 1f, 1f - tp1.center.y * 2f)
                }
                i == n - 1 -> {
                    val (tpP, _) = pts[n - 2]
                    GLHelper.segNormal(tpP.center.x * 2f - 1f, 1f - tpP.center.y * 2f, x, y)
                }
                else -> {
                    val (tpP, _) = pts[i - 1]; val (tpN, _) = pts[i + 1]
                    GLHelper.avgNormal(
                        tpP.center.x * 2f - 1f, 1f - tpP.center.y * 2f,
                        x, y,
                        tpN.center.x * 2f - 1f, 1f - tpN.center.y * 2f
                    )
                }
            }

            val hw = halfWidth * alpha * (1f - alpha * 0.7f)
            val c = headColor
            fun ch(base: Float) = (base + (1f - base) * coreBoost * alpha).coerceIn(0f, 1f)
            val r = ch((c shr 16 and 0xFF) / 255f)
            val g = ch((c shr  8 and 0xFF) / 255f)
            val b = ch((c        and 0xFF) / 255f)
            val a = alpha * alphaScale

            ctx.ribbonFloats.put(x - nx * hw).put(y - ny * hw).put(r).put(g).put(b).put(a)
            ctx.ribbonFloats.put(x + nx * hw).put(y + ny * hw).put(r).put(g).put(b).put(a)
        }

        ctx.ribbonBuf.position(0)
        GLES20.glVertexAttribPointer(posLoc,   2, GLES20.GL_FLOAT, false, BYTES_PER_VERT, ctx.ribbonBuf)
        ctx.ribbonBuf.position(8)
        GLES20.glVertexAttribPointer(colorLoc, 4, GLES20.GL_FLOAT, false, BYTES_PER_VERT, ctx.ribbonBuf)
        ctx.ribbonBuf.position(0)
        GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, 0, n * 2)
    }

    companion object {
        private const val BYTES_PER_VERT = 24  // 6 floats × 4 bytes

        private val TRAIL_VERT = """
            attribute vec2 aPosition;
            attribute vec4 aColor;
            varying vec4 vColor;
            void main() {
                gl_Position = vec4(aPosition, 0.0, 1.0);
                vColor = aColor;
            }""".trimIndent()

        private val TRAIL_FRAG = """
            precision mediump float;
            varying vec4 vColor;
            void main() {
                gl_FragColor = vColor;
            }""".trimIndent()
    }
}

File: tmp/metal_effect_backup_20260630_140225/UI/GLEffects/GenericGLEffect.swift
import OpenGLES
import UIKit

/// Direct port of GenericGLEffect.kt.
final class GenericGLEffect: GLEffect {
  private var program: GLuint = 0
  private var posLoc: GLint = -1
  private var colorLoc: GLint = -1

  override func onGLReady(context: GLRenderContext) {
    program = GLHelper.buildProgram(
      vertexSource: GLEffectShaders.Generic.trailVert,
      fragmentSource: GLEffectShaders.Generic.trailFrag
    )
    posLoc = glGetAttribLocation(program, "aPosition")
    colorLoc = glGetAttribLocation(program, "aColor")
  }

  override func draw(
    trackData: GLTrackData,
    context: GLRenderContext,
    effectType: EffectType
  ) {
    glUseProgram(program)
    for (_, points) in trackData where points.count >= 2 {
      let color = effectType.colorOverride ?? points.last!.first.color
      drawRibbon(
        points,
        halfWidth: 0.070 * effectType.glowWidthMult,
        alphaScale: 0.45,
        coreBoost: 0,
        headColor: color,
        context: context
      )
      drawRibbon(
        points,
        halfWidth: 0.022 * effectType.coreWidthMult,
        alphaScale: 0.92,
        coreBoost: 0.55,
        headColor: color,
        context: context
      )
    }
  }

  private func drawRibbon(
    _ points: [GLTrailSample],
    halfWidth: Float,
    alphaScale: Float,
    coreBoost: Float,
    headColor: UIColor,
    context: GLRenderContext
  ) {
    let n = points.count
    guard n >= 2,
      n * 2 * Self.bytesPerVertex <= context.ribbonFloats.capacityBytes
    else {
      return
    }

    let base = GLHelper.rgba(headColor)
    context.ribbonFloats.clear()

    for i in 0..<n {
      let sample = points[i]
      let point = sample.first
      let alpha = sample.second
      let x = Float(point.center.x * 2 - 1) * context.quadScaleX
      let y = Float(1 - point.center.y * 2) * context.quadScaleY

      let normal: (Float, Float)
      if i == 0 {
        let next = points[1].first.center
        normal = GLHelper.segNormal(
          x, y,
          Float(next.x * 2 - 1) * context.quadScaleX,
          Float(1 - next.y * 2) * context.quadScaleY
        )
      } else if i == n - 1 {
        let previous = points[n - 2].first.center
        normal = GLHelper.segNormal(
          Float(previous.x * 2 - 1) * context.quadScaleX,
          Float(1 - previous.y * 2) * context.quadScaleY,
          x, y
        )
      } else {
        let previous = points[i - 1].first.center
        let next = points[i + 1].first.center
        normal = GLHelper.avgNormal(
          Float(previous.x * 2 - 1) * context.quadScaleX,
          Float(1 - previous.y * 2) * context.quadScaleY,
          x, y,
          Float(next.x * 2 - 1) * context.quadScaleX,
          Float(1 - next.y * 2) * context.quadScaleY
        )
      }

      let half = halfWidth * alpha * (1 - alpha * 0.7)
      let brighten: (Float) -> Float = { component in
        (component + (1 - component) * coreBoost * alpha).glClamped()
      }
      let r = brighten(base.0)
      let g = brighten(base.1)
      let b = brighten(base.2)
      let a = alpha * alphaScale

      context.ribbonFloats
        .put(x - normal.0 * half).put(y - normal.1 * half)
        .put(r).put(g).put(b).put(a)
      context.ribbonFloats
        .put(x + normal.0 * half).put(y + normal.1 * half)
        .put(r).put(g).put(b).put(a)
    }

    GLHelper.drawInterleaved(
      buffer: context.ribbonFloats,
      strideBytes: Self.bytesPerVertex,
      attributes: [
        GLVertexAttribute(location: posLoc, size: 2, offsetBytes: 0),
        GLVertexAttribute(location: colorLoc, size: 4, offsetBytes: 8),
      ],
      mode: GLenum(GL_TRIANGLE_STRIP),
      vertexCount: n * 2
    )
  }

  private static let bytesPerVertex = 24
}
